[DTensor] Fix _StridedShard to Replicate padding issue

zpcore opened the pull request:
Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.

See below picture for illustration on how the StridedShard data get reshuffled back after allgather (mesh dim 0 size is 4) :
<img width="671" height="724" alt="image" src="https://github.com/user-attachments/assets/9c0e10fc-a0e4-41fe-bae6-1017ff660e51" />
code accompanying diagram for index select usage:
```
        padded_positions = []
        for i, shard in enumerate(sharded_indices):
            base_offset = i * max_chunk_size
            positions = base_offset + torch.arange(
                len(shard), device=local_tensor.device
            )
            padded_positions.append(positions)

        permutation = torch.cat(sharded_indices)
        # Choose the position by skipping padding indices from
        # replicate_tensor_permuted_padded.
        select_positions = torch.cat(padded_positions)

        inv_permutation = torch.argsort(permutation)
        select_indices = select_positions.index_select(0, inv_permutation)

        replicate_tensor = torch.index_select(
            replicate_tensor_permuted_padded, self.dim, select_indices
        )

```
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171338
* #171219
* #171178
* #171023
* #170915
* #172266
* #171337
* __->__ #170914

zpcore committed:
[DTensor] Fix _StridedShard to Replicate padding issue

[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.






[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.






[ghstack-poisoned]

wconstab reviewed (commented):
can you update the PR desc to show 2 things

1) tell us what the mesh dim size is
2) run python interpreter to simulate the torch.chunk behavior directly? IIUC the way StridedShard got created via _split_tensor, it should call chunk twice.  But i am not sure I understand how this results in the padding combination you claimed.

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.






[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.

See below picture for illustration (mesh dim 0 size is 4):
<img width="834" height="661" alt="image" src="https://github.com/user-attachments/assets/58de06e3-be56-4a3b-9c26-1d3d1b08c2fe" />






[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.

See below picture for illustration on how the StridedShard data get reshuffled back after allgather (mesh dim 0 size is 4) :
<img width="671" height="724" alt="image" src="https://github.com/user-attachments/assets/9c0e10fc-a0e4-41fe-bae6-1017ff660e51" />




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.

See below picture for illustration on how the StridedShard data get reshuffled back after allgather (mesh dim 0 size is 4) :
<img width="671" height="724" alt="image" src="https://github.com/user-attachments/assets/9c0e10fc-a0e4-41fe-bae6-1017ff660e51" />




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.

See below picture for illustration on how the StridedShard data get reshuffled back after allgather (mesh dim 0 size is 4) :
<img width="671" height="724" alt="image" src="https://github.com/user-attachments/assets/9c0e10fc-a0e4-41fe-bae6-1017ff660e51" />




[ghstack-poisoned]

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.

See below picture for illustration on how the StridedShard data get reshuffled back after allgather (mesh dim 0 size is 4) :
<img width="671" height="724" alt="image" src="https://github.com/user-attachments/assets/9c0e10fc-a0e4-41fe-bae6-1017ff660e51" />




[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -803,19 +804,30 @@ def _to_replicate_tensor(
+            padded_positions.append(positions)
 
         permutation = torch.cat(sharded_indices)
+        # Choose the position by skipping padding indices from
```
is torch.cat(sharded_indices) always equivalent to torch.arange(n) or is there some reason it is not? i think it would be clearer to just use arange if so, it is difficult to trace this code

wconstab reviewed (commented):
lgtm. it took me a while to understand the code though..

weifengpy commented on test/distributed/tensor/test_utils.py:
```diff
@@ -1496,6 +1498,124 @@ def tearDown(self):
+
+    ShardConfig = namedtuple("ShardConfig", ["mesh_dim", "split_factor"], defaults=(1,))
+
+    def _range_tensor(self, *shape) -> torch.Tensor:
```
who calls this `_range_tensor`?

zpcore replied:
Will be used in future PRs.  I have relocated it to that PR.

zpcore committed:
Update on "[DTensor] Fix _StridedShard to Replicate padding issue"

Fix `_StridedShard._to_replicate_tensor` incorrectly unpad the tensors after all gather. 

`_StridedShard._to_replicate_tensor` previously assumes the padding can only happen at the end of the gathered tensor. However, due to the split_factor, the padding can occur on different chunks. E.g., 

```
import torch
from torch.distributed.tensor.placement_types import _StridedShard
a = torch.arange(9) # 1D tensor with 9 elements
SS = _StridedShard(0, split_factor=2)
num_chunks = 4
output = SS._split_tensor(a, num_chunks)
```
We can see the output splitted tensor contains 4 chunks:
`[0, 1, 5], [2, 3, 6], [4, 7], [8]`
When we try pad before allgather, the 4 chunks become:
`[0, 1, 5], [2, 3, 6], [4, 7, Pad], [8, Pad, Pad]`
After allgather, we have:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad, 8, Pad, Pad]`
Original assumption is that there are total 3 paddings (based on max_chunk_size * num_splits), so we trim the last 3 elements, and the final result becomes:
`[0, 1, 5, 2, 3, 6, 4, 7, Pad]`
Now we keep track of the padding and will only select the nonpadding elements after allgather.

See below picture for illustration on how the StridedShard data get reshuffled back after allgather (mesh dim 0 size is 4) :
<img width="671" height="724" alt="image" src="https://github.com/user-attachments/assets/9c0e10fc-a0e4-41fe-bae6-1017ff660e51" />
code accompanying diagram for index select usage:
```
        padded_positions = []
        for i, shard in enumerate(sharded_indices):
            base_offset = i * max_chunk_size
            positions = base_offset + torch.arange(
                len(shard), device=local_tensor.device
            )
            padded_positions.append(positions)

        permutation = torch.cat(sharded_indices)
        # Choose the position by skipping padding indices from
        # replicate_tensor_permuted_padded.
        select_positions = torch.cat(padded_positions)

        inv_permutation = torch.argsort(permutation)
        select_indices = select_positions.index_select(0, inv_permutation)

        replicate_tensor = torch.index_select(
            replicate_tensor_permuted_padded, self.dim, select_indices
        )

```



[ghstack-poisoned]

wconstab reviewed (approved):
lgtm!

zpcore commented:
@pytorchbot merge
